controllers/snapshot/snapshot_adapter.go

```diff
@@ -0,0 +1,444 @@
+	opts := []client.ListOption{
+		client.InNamespace(applicationSnapshot.Namespace),
+		client.MatchingFields{"spec.applicationSnapshot": applicationSnapshot.Name},
+	}
+
+	err := a.client.List(a.context, releases, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &releases.Items, nil
+}
+
+// createReleaseForReleasePlan creates the Release for a given ReleasePlan.
+// In case the Release can't be created, an error will be returned.
+func (a *Adapter) createReleaseForReleasePlan(releasePlan *releasev1alpha1.ReleasePlan, applicationSnapshot *appstudioshared.ApplicationSnapshot) (*releasev1alpha1.Release, error) {
+	newRelease := &releasev1alpha1.Release{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: applicationSnapshot.Name + "-",
+			Namespace:    applicationSnapshot.Namespace,
+		},
+		Spec: releasev1alpha1.ReleaseSpec{
+			ApplicationSnapshot: applicationSnapshot.Name,
+			ReleasePlan:         releasePlan.Name,
+		},
+	}
+	err := a.client.Create(a.context, newRelease)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRelease, nil
+}
+
+// createMissingReleasesForReleasePlans checks if there's existing Releases for a given list of ReleasePlans and creates
+// new ones if they are missing. In case the Releases can't be created, an error will be returned.
+func (a *Adapter) createMissingReleasesForReleasePlans(releasePlans *[]releasev1alpha1.ReleasePlan, applicationSnapshot *appstudioshared.ApplicationSnapshot) error {
+	releases, err := a.getReleasesWithApplicationSnapshot(applicationSnapshot)
+	if err != nil {
+		return err
+	}
+
+	for _, releasePlan := range *releasePlans {
+		releasePlan := releasePlan
+		var existingRelease *releasev1alpha1.Release = nil
+		for _, snapshotRelease := range *releases {
+			snapshotRelease := snapshotRelease
+			if snapshotRelease.Spec.ReleasePlan == releasePlan.Name {
+				existingRelease = &snapshotRelease
+			}
+		}
+		if existingRelease != nil {
+			a.logger.Info("Found existing Release",
+				"ApplicationSnapshot.Name", applicationSnapshot.Name,
+				"ReleasePlan.Name", releasePlan.Name,
+				"Release.Name", existingRelease.Name)
+		} else {
+			newRelease, err := a.createReleaseForReleasePlan(&releasePlan, applicationSnapshot)
+			if err != nil {
+				return err
+			}
+			a.logger.Info("Created new Release",
+				"Application.Name", a.application.Name,
+				"ReleasePlan.Name", releasePlan.Name,
+				"Release.Name", newRelease.Name)
+		}
+	}
+	return nil
+}
+
+// getRequiredIntegrationTestScenariosForApplication returns the IntegrationTestScenarios used by the application being processed.
+// A IntegrationTestScenarios will only be returned if it has the
+// release.appstudio.openshift.io/optional label set to true or if it is missing the label entirely.
+func (a *Adapter) getRequiredIntegrationTestScenariosForApplication(application *hasv1alpha1.Application) (*[]v1alpha1.IntegrationTestScenario, error) {
+	labelSelector := labels.NewSelector()
```

Comment:
This line is not needed. Just use a colon during the assignment.

Reply:
I think it makes sense here because the next call for it is adding a requirement to an empty selector instead of assigning it. It's the same case as in the [getAutoReleasePlansForApplication](https://github.com/redhat-appstudio/integration-service/blob/main/controllers/integration/integration_adapter.go#L578).

Reply:
Yeah, not sure what I was thinking about when I wrote this. I actually remember my thoughts when I reviewed initially and what I wanted to write is that the commented line should be moved down to the relevant part (just before you adding the requirement). Alternatively you could write it like: `labelSelector := labels.NewSelector().Add(*labelRequirement)` given that there's no error checking needed.

Reply:
Yeah, that makes it a bit more compact. Updated the [two](https://github.com/redhat-appstudio/integration-service/pull/29/files#diff-ea27f7cada77e890e9c4d991d8edc8f9c6bfbbf2a830cf74df4ba606341240caR39) [occurrences](https://github.com/redhat-appstudio/integration-service/pull/29/files#diff-c311e0da05672ddb12ae96960655916c29bb57773e495aece9db22a54d04b28dR66) to use that approach.

Reply:
Resolving to unblock the subsequent PRs